test/integration/cds_integration_test.cc

```diff
@@ -222,6 +222,122 @@ TEST_P(CdsIntegrationTest, CdsClusterTeardownWhileConnecting) {
+      {":path", "/cluster1"},
+      {":scheme", "http"},
+      {":authority", "host"},
+  };
+  BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
+      lookupPort("http"), "GET", "/cluster1", "", downstream_protocol_, version_, "foo.com");
+  ASSERT_TRUE(response->complete());
+  cleanupUpstreamAndDownstream();
+
+  // Cluster_1 trafficStats updated.
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_1.ClusterTrafficStats.inited")->value(), 1);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total")->value(), 1);
+  envoy::config::cluster::v3::Cluster cluster1_updated =
+      cluster_creator_(ClusterName1, fake_upstreams_[UpstreamIndex2]->localAddress()->ip()->port(),
+                       Network::Test::getLoopbackAddressString(ipVersion()), "ROUND_ROBIN");
+  sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(
+      Config::TypeUrl::get().Cluster, {cluster1_updated}, {cluster1_updated}, {}, "42");
+  test_server_->waitForCounterGe("cluster_manager.cds.update_success", 2);
+  // Now the ClusterTrafficStats 'inited_' gauge >= 1, since there is a small race window between
+  // the testing thread fetching the value before the old version ClusterTrafficStats gets deleted
+  // from main thread and worker thread.
+  EXPECT_GE(test_server_->gauge("cluster.cluster_1.ClusterTrafficStats.inited")->value(), 1);
+  // cluster_1 traffic stats not lost.
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total")->value(), 1);
+
+  // Remove "cluster_1".
+  sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(Config::TypeUrl::get().Cluster, {}, {},
+                                                             {ClusterName1}, "42");
+  test_server_->waitForCounterGe("cluster_manager.cluster_removed", 1);
+  test_server_->waitForCounterGe("cluster_manager.cds.update_success", 3);
+  sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(Config::TypeUrl::get().Cluster,
+                                                             {cluster2_}, {cluster2_}, {}, "43");
+  test_server_->waitForCounterGe("cluster_manager.cds.update_success", 4);
+  EXPECT_EQ(test_server_->counter("cluster_manager.cluster_added")->value(), 3);
+  // Now the cluster_1 stats are gone, as well as the lazy init wrapper.
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_1.ClusterTrafficStats.inited"), nullptr);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total"), nullptr);
+  // No cluster_2 stats yet.
+  EXPECT_NE(test_server_->gauge("cluster.cluster_2.ClusterTrafficStats.inited"), nullptr);
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_2.ClusterTrafficStats.inited")->value(), 0);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_2.upstream_cx_total"), nullptr);
+}
+
+// Test that LazyInitTrafficStats with cluster_1 create-remove-create sequence.
+TEST_P(CdsIntegrationTest, LazyInitTrafficStatsWithClusterCreateDeleteRecrete) {
+  this->use_real_stats_ = true;
```

Comment:
LInes 227-235 are the same as lines 283-291.

I think you should subclass CdsIntegrationTest (say LazyCdsStatsIntegrationTest) to put some shared methods to create a bunch of smaller easy-to-read tests for lazy funcitonality.

Reply:
done